### ISSUE TITLE
FAILED PersistenceEventAdapterDocSpec

### ISSUE THREAD

**Issue by 2m** [akka/akka#20400]
While validating https://github.com/akka/akka/pull/20397

```
[info] MyAutomaticJsonEventAdapter
[info] - must demonstrate how to implement a JSON adapter *** FAILED *** (7 seconds, 104 milliseconds)
[info]   java.lang.AssertionError: assertion failed: timeout (3 seconds) during expectMsg while waiting for Person(Caplin,42)
[info]   at scala.Predef$.assert(Predef.scala:170)
[info]   at akka.testkit.TestKitBase$class.expectMsg_internal(TestKit.scala:349)
[info]   at akka.testkit.TestKitBase$class.expectMsg(TestKit.scala:326)
[info]   at akka.testkit.TestKit.expectMsg(TestKit.scala:734)
[info]   at docs.persistence.PersistenceEventAdapterDocSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(PersistenceEventAdapterDocSpec.scala:93)
[info]   at docs.persistence.PersistenceEventAdapterDocSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(PersistenceEventAdapterDocSpec.scala:71)
[info]   at docs.persistence.PersistenceEventAdapterDocSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(PersistenceEventAdapterDocSpec.scala:71)
[info]   at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
...
```

https://jenkins.akka.io:8498/job/pr-validator-per-commit-jenkins/5971/consoleFull

**Comment by patriknw** [comment]
The test looks correct and it's using in-mem journal so 3 seconds should be enough. Let's see if it happens again.

**Comment by patriknw** [comment]
https://jenkins.akka.io:8498/job/pr-validator-per-commit-jenkins/6562/consoleFull

**Comment by johanandren** [comment]
It is doing a 8.0292939 gc pause just before or while running, so given that 3s doesn't help much?

**Comment by patriknw** [comment]
yeah, wow

```
5280.174: [Full GC (Ergonomics) [PSYoungGen: 571904K->115820K(635392K)] [ParOldGen: 1397761K->1398264K(1398272K)] 1969665K->1514084K(2033664K), [Metaspace: 587058K->587058K(1548288K)], 8.0292939 secs] [Times: user=224.76 sys=1.58, real=8.03 secs] 
```

The first failure 5971 didn't have any such gc logging, though.

I think we close this anyway

**Comment by johanandren** [comment]
Another one: https://jenkins.akka.io:8498/job/akka-nightly/3739/consoleFull

```
[info] - must demonstrate how to implement a JSON adapter *** FAILED *** (3 seconds, 199 milliseconds)
[info]   java.lang.AssertionError: assertion failed: timeout (3 seconds) during expectMsg while waiting for Person(Caplin,42)
[info]   at scala.Predef$.assert(Predef.scala:170)
[info]   at akka.testkit.TestKitBase$class.expectMsg_internal(TestKit.scala:387)
[info]   at akka.testkit.TestKitBase$class.expectMsg(TestKit.scala:364)
[info]   at akka.testkit.TestKit.expectMsg(TestKit.scala:814)
[info]   at docs.persistence.PersistenceEventAdapterDocSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(PersistenceEventAdapterDocSpec.scala:93)
[info]   at docs.persistence.PersistenceEventAdapterDocSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(PersistenceEventAdapterDocSpec.scala:71)
```

**Comment by johanandren** [comment]
no GC pauses for the one Oct 3

**Comment by drewhk** [comment]
Quite old, but since it seems to happen rather rarely, I keep this open for a while.

**Comment by ktoso** [comment]
This seems to not have happened since January, closing?

**Comment by raboof** [comment]
Unfortunately failed again, again after a 4-second GC pause: https://jenkins.akka.io:8498/job/akka-multi-node-nightly2/756/consoleFull
```
[info] MyAutomaticJsonEventAdapter
5970.974: [Full GC (Metadata GC Threshold)  1273M->578M(3072M), 4.0180459 secs]
   [Eden: 175.0M(1597.0M)->0.0B(1843.0M) Survivors: 47.0M->0.0B Heap: 1273.3M(3072.0M)->578.5M(3072.0M)], [Metaspace: 502683K->479917K(1492992K)]
 [Times: user=5.62 sys=0.13, real=4.01 secs] 
[info] - must demonstrate how to implement a JSON adapter *** FAILED *** (4 seconds, 47 milliseconds)
[info]   java.lang.AssertionError: assertion failed: timeout (3 seconds) during expectMsg while waiting for Person(Caplin,42)
[info]   at scala.Predef$.assert(Predef.scala:170)
[info]   at akka.testkit.TestKitBase$class.expectMsg_internal(TestKit.scala:386)
[info]   at akka.testkit.TestKitBase$class.expectMsg(TestKit.scala:363)
[info]   at akka.testkit.TestKit.expectMsg(TestKit.scala:831)
[info]   at docs.persistence.PersistenceEventAdapterDocSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(PersistenceEventAdapterDocSpec.scala:93)
[info]   at docs.persistence.PersistenceEventAdapterDocSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(PersistenceEventAdapterDocSpec.scala:71)
```

**Comment by raboof** [comment]
And another one, again after a slow GC
```
5963.277: [Full GC (Metadata GC Threshold)  1724M->598M(3072M), 4.2474760 secs]
   [Eden: 669.0M(1627.0M)->0.0B(1843.0M) Survivors: 60.0M->0.0B Heap: 1724.0M(3072.0M)->598.6M(3072.0M)], [Metaspace: 504222K->482324K(1488896K)]
 [Times: user=7.63 sys=0.13, real=4.25 secs] 
[info] - must demonstrate how to implement a JSON adapter *** FAILED *** (4 seconds, 269 milliseconds)
[info]   java.lang.AssertionError: assertion failed: timeout (3 seconds) during expectMsg while waiting for Person(Caplin,42)
[info]   at scala.Predef$.assert(Predef.scala:170)
[info]   at akka.testkit.TestKitBase$class.expectMsg_internal(TestKit.scala:386)
[info]   at akka.testkit.TestKitBase$class.expectMsg(TestKit.scala:363)
[info]   at akka.testkit.TestKit.expectMsg(TestKit.scala:831)
[info]   at docs.persistence.PersistenceEventAdapterDocSpec$$anonfun$2$$anonfun$apply$mcV$sp$2.apply(PersistenceEventAdapterDocSpec.scala:125)
[info]   at docs.persistence.PersistenceEventAdapterDocSpec$$anonfun$2$$anonfun$apply$mcV$sp$2.apply(PersistenceEventAdapterDocSpec.scala:103)
```

**Comment by patriknw** [comment]
So it's again the metaspace that is causing trouble?
Metadata GC Threshold

**Comment by patriknw** [comment]
I see two options.

1) We increase the `-XX:MetaspaceSize` even more so that we don't hit a full GC due to metaspace, https://stackoverflow.com/questions/25251388/what-is-the-metadata-gc-threshold-and-how-do-i-tune-it
2) We try to fork tests. Would be nice with one fork per project. http://www.scala-sbt.org/0.13/docs/Testing.html#Forking+tests

**Comment by raboof** [comment]
Hmm, while it's a metaspace resize that triggers the Full GC, isn't the problem that the GC takes so long, rather than that it is triggered? Or are we aiming to prevent Full GC's at all?

Nonetheless it might be nice to fork the tests, though, as that'd also allow them to run in parallel (and potentially finish faster), right?

**Comment by patriknw** [comment]
Unless it's the actual metaspace gc that takes time. I don't think metaspace is supposed to be this large. Default is something lik 16M. This build is not like an ordinary jvm app.

**Comment by johanandren** [comment]
I also wonder why we use so much metaspace, do we have a class-loader leak perhaps? Some class getting registered to a singleton keeping it alive across all tests?

**Comment by johanandren** [comment]
Hopefully fixed by #23771
